src/systems/userinput/userinput.js

```diff
@@ -157,21 +158,51 @@ function computeExecutionStrategy(sortedBindings, masks, activeSets) {
 
 AFRAME.registerSystem("userinput", {
   get(path) {
-    return this.frame && this.frame[path];
+    if (!this.frame) return;
+    return this.frame.get(path);
   },
 
   toggleSet(set, value) {
     this.pendingSetChanges.push({ set, value });
   },
 
   init() {
-    this.frame = {};
+    this.frame = {
+      generation: 0,
+      values: {},
+      generations: {},
+      get: function(path) {
+        if (this.generations[path] !== this.generation) return undefined;
+        return this.values[path];
+      },
+      setValueType: function(path, value) {
+        this.values[path] = value;
+        this.generations[path] = this.generation;
+      },
+      setVector2: function(path, a, b) {
+        const value = this.values[path] || [];
+        value[0] = a;
+        value[1] = b;
+        this.values[path] = value;
+        this.generations[path] = this.generation;
+      },
+      setPose: function(path, pose) {
+        this.setValueType(path, pose);
+      },
+      setMatrix4: function(path, mat4) {
+        // Should we assume the incoming mat4 is safe to store instead of copying values?
+        const value = this.values[path] || new THREE.Matrix4();
+        value.copy(mat4);
+        this.values[path] = value;
+        this.generations[path] = this.generation;
+      }
+    };
 
     this.prevActiveSets = new Set();
     this.activeSets = new Set([sets.global]);
     this.pendingSetChanges = [];
     this.xformStates = new Map();
-    this.activeDevices = new Set([new HudDevice()]);
+    this.activeDevices = new ArrayBackedSet().add(new HudDevice());
```

Comment:
Instead of doing this chaining I would just allow the constructor to take an array to be used as the backing array, and pass in the value like the `Set` version.

Reply:
Oh that makes the most sense